### PR TITLE
RESTful storage prototype.

### PR DESCRIPTION
**Pull request by mzhurovich**
Hi @dkorolev!
Here is the prototype of the RESTful layer for our storage.
It supports only `Dictionary` in a pretty simple manner, but workz ;)
What is needed to be discussed:
* Some transparent way to return more sophisticated results (Hypermedia etc.).
* Painless injection of custom user code into handlers (authentication etc.). Handlers chaining?

### CODE REVIEW

File: Storage/test.cc [hunk 1]
```diff
@@ -48,7 +54,7 @@ namespace transactional_storage_test {
 CURRENT_STRUCT(Element) {
   CURRENT_FIELD(x, int32_t);
   CURRENT_CONSTRUCTOR(Element)(int32_t x = 0) : x(x) {}
-  // Dummy key function to support key-only deleter.
+  // To use `Element` in storage, it should either define a `.key` field, or `.key()` method.
```
Comment (dkorolev) [inline Storage/test.cc]: This is actually a bug/feature: `Element` is only used in `Vector`, which doesn't need the `key/key()`. We'll take care of it soon-ish.

File: Storage/rest.h [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+  static void Run(HTTPRoutesScope&, T_SERVER&, T_STORAGE&) {}
+};
+
+}  // namespace impl
+
+template <class T_STORAGE_IMPL>
+struct RESTfulStorage {
+  HTTPRoutesScope handlers_scope;
+  RESTfulStorage(T_STORAGE_IMPL& storage, int port) {
+    auto& http_server = HTTP(port);
```
Comment (dkorolev) [inline Storage/rest.h]: :-)

File: Storage/rest.h [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+  template <typename T_SERVER, typename T_STORAGE>
+  static void Run(HTTPRoutesScope& scope, T_SERVER& http_server, T_STORAGE& storage) {
+    scope += RegisterRESTfulStorageEndpoint<N>(http_server, storage);
+    RegisterRESTfulStorageEndpoints<current::variadic_indexes::indexes<NS...>>::Run(
+        scope, http_server, storage);
+  }
+};
+
+template <>
+struct RegisterRESTfulStorageEndpoints<current::variadic_indexes::indexes<>> {
```
Comment (dkorolev) [inline Storage/rest.h]: Looks like a simple `template<int I, int N>` calling `<I+1, N>`, and having a corner case `<N, N>` would do here, no need in `variadic_indexes`.

File: Storage/test.cc [hunk 4]
```diff
@@ -569,5 +613,147 @@ TEST(TransactionalStorage, OnDisk) {
...
+    EXPECT_EQ("MZ", ParseJSON<SimpleUser>(HTTP(GET(base_url + "/user/max")).body).name);
+
+    EXPECT_EQ(204, static_cast<int>(HTTP(DELETE(base_url + "/user/max")).code));
+    EXPECT_EQ(404, static_cast<int>(HTTP(GET(base_url + "/user/max")).code));
+  }
+}
+
+// TODO: `::current::storage::FieldNameAndTypeByIndex<0>()` can extract storage entry type ("Dictionary").
+
+// A "complete" "REST API" test.
```
Comment (dkorolev) [inline Storage/test.cc]: Why quotes? It's indeed a REST API and a test for it! :-)

File: Storage/test.cc [hunk 5]
```diff
@@ -569,5 +613,147 @@ TEST(TransactionalStorage, OnDisk) {
...
+  const auto persistence_file_remover = current::FileSystem::ScopedRmFile(persistence_file_name);
+
+  Storage storage(persistence_file_name);
+  EXPECT_EQ(2u, storage.FieldsCount());
+
+  const auto base_url = current::strings::Printf("http://localhost:%d", FLAGS_transactional_storage_test_port);
+
+  // Run twice to make sure the `GET-POST-GET-DELETE` cycle is complete.
+  for (size_t i = 0; i < 2; ++i) {
+    // Registed RESTful HTTP endpoints, in a scoped way.
```
Comment (dkorolev) [inline Storage/test.cc]: s/Registed/Register/

File: TypeSystem/Serialization/types/variant.h [hunk 6]
```diff
@@ -30,9 +30,7 @@ SOFTWARE.
 
 #include "primitive.h"
 
-// To fix `error: incomplete type ‘current::serialization::json::load::LoadFromJSONImpl<>`.
-#include "../json.h"
-#include "../binary.h"
+#include "current_typeid.h"
```
Comment (dkorolev) [inline TypeSystem/Serialization/types/variant.h]: :+1:

**Comment by dkorolev** [comment]
(LGTM++)++!
